pkg/apis/pipeline/v1beta1/pipeline_validation_test.go

```diff
@@ -1865,6 +1883,28 @@ func TestValidatePipelineWithFinalTasks_Failure(t *testing.T) {
 			},
 		},
 		expectedError: *apis.ErrGeneric("expected at least one, got none", "spec.description", "spec.params", "spec.resources", "spec.tasks", "spec.workspaces"),
+	}, {
+		name: "invalid pipeline with final tasks referring to invalid context variables",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Tasks: []PipelineTask{{
+					Name:    "non-final-task",
+					TaskRef: &TaskRef{Name: "non-final-task"},
+				}},
+				Finally: []PipelineTask{{
+					Name:    "final-task-1",
+					TaskRef: &TaskRef{Name: "final-task"},
+					Params: []Param{{
+						Name: "param1", Value: ArrayOrString{Type: ParamTypeString, StringVal: "$(context.pipelineRun.missing)"},
+					}},
+				}},
+			},
+		},
+		expectedError: apis.FieldError{
+			Message: `non-existent variable in "$(context.pipelineRun.missing)"`,
```

Comment:
not related to this PR but this message is not accurate and we can improve on it (applies to both `tasks` and `finally`). This is getting generated from the `substitution.go` and very generic.

Reply:
agreed, we can improve the error messages in another PR